TST use global_random_seed in sklearn/ensemble/tests/test_iforest.py

MaxwellLZH opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Related issue #22827.

#### What does this implement/fix? Explain your changes.
Adds `global_random_seed` to tests that are checking model precision or consistency. 

The only tests that didn't work out-of-the box was `test_iforest_performance` which fails on a couple of seeds, so I increased sample size and made the outlier data a bit more obvious.


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

MaxwellLZH committed:
add global-random-seed in tests

jeremiedbb commented on sklearn/ensemble/tests/test_iforest.py:
```diff
@@ -165,19 +165,19 @@ def test_iforest_parallel_regression():
+    X = 0.3 * rng.randn(1200, 2)
     X_train = np.r_[X + 2, X - 2]
-    X_train = X[:100]
+    X_train = X[:1000]
 
     # Generate some abnormal novel observations
-    X_outliers = rng.uniform(low=-4, high=4, size=(20, 2))
-    X_test = np.r_[X[100:], X_outliers]
-    y_test = np.array([0] * 20 + [1] * 20)
+    X_outliers = rng.uniform(low=-16, high=16, size=(200, 2))
+    X_test = np.r_[X[1000:], X_outliers]
+    y_test = np.array([0] * 200 + [1] * 200)
```
This test is already not tested what it's supposed to.
```py
X = 0.3 * rng.randn(120, 2)  # generate 120 points centered at the origin
X_train = np.r_[X + 2, X - 2]  # duplicate such that 120 points centered at (-2,-2) and 120 at (2,2)
X_train = X[:100]  # ignore the line just above and take 100 points centered at the origin
```
Then same for X_test...

The test is also brittle. When testing multiple seeds it often happen that the X_test taken from X are not mixed with the ones of X_train and thus are misclassified.
I propose to modify the test as follows, with a little bit less samples than you proposed
```py
    # Generate train/test data
    rng = np.random.RandomState(global_random_seed)
    X = 0.3 * rng.randn(600, 2)
    X = rng.permutation(np.vstack((X + 2, X - 2)))
    X_train = X[:1000]

    # Generate some abnormal novel observations
    X_outliers = rng.uniform(low=-1, high=1, size=(200, 2))
    X_test = np.vstack((X[1000:], X_outliers))
    y_test = np.array([0] * 200 + [1] * 200)
```

MaxwellLZH replied:
Hi @jeremiedbb, I've updated the test case as you suggested, and verifies that it works for all the random seeds :)

jeremiedbb commented on sklearn/ensemble/tests/test_iforest.py:
```diff
@@ -303,15 +303,15 @@ def test_iforest_chunks_works2(mocked_get_chunk, contamination, n_predict_calls)
 
 
-def test_iforest_with_uniform_data():
+def test_iforest_with_uniform_data(global_random_seed):
```
I don't think this test should be parametrized. It's just sanity checks

jeremiedbb reviewed (commented):
Thanks for the PR @MaxwellLZH. Here are some suggestions

MaxwellLZH committed:
remove global_random_seed for test_iforest_with_uniform_data

MaxwellLZH committed:
update test_iforest_performance

jeremiedbb reviewed (approved):
LGTM. Thanks @MaxwellLZH

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/MaxwellLZH/22901

jeremiedbb committed:
[all random seeds]
test_iforest_sparse
test_iforest_parallel_regression
test_iforest_performance

jeremiedbb commented:
tested with all random seeds in commit [768d4f2](https://github.com/scikit-learn/scikit-learn/pull/22901/commits/768d4f24be32d6533ca546d2f8d75d290dd52093)

jeremiedbb commented on sklearn/ensemble/tests/test_iforest.py:
```diff
@@ -58,9 +58,9 @@ def test_iforest():
 
 
-def test_iforest_sparse():
+def test_iforest_sparse(global_random_seed):
```
In this test, IsolationForest is still instanciated with random_state=1. I think it would make sense to set `random_state=global_random_seed`

jeremiedbb commented on sklearn/ensemble/tests/test_iforest.py:
```diff
@@ -143,9 +143,9 @@ def test_max_samples_attribute():
 
 
-def test_iforest_parallel_regression():
+def test_iforest_parallel_regression(global_random_seed):
```
In this test, IsolationForest is still instanciated with random_state=0. I think it would make sense to set `random_state=global_random_seed`

MaxwellLZH committed:
use global_random_seed in iforest construction

glemaitre commented on sklearn/ensemble/tests/test_iforest.py:
```diff
@@ -58,9 +58,9 @@ def test_iforest():
             IsolationForest(random_state=rng, **params).fit(X_train).predict(X_test)
```
I assume that we also want the global random state here as well.

glemaitre replied:
I think that we should remove the line 30

```python
rng = np.random.RandomState(0)
```

We should not share the same random state in the tests. Instead we should locally create the random state internally in the function or instead use the global random state.

MaxwellLZH replied:
I've updated the tests so that the global random state is now only used to permute the dataset.

glemaitre commented on sklearn/ensemble/tests/test_iforest.py:
```diff
@@ -58,9 +58,9 @@ def test_iforest():
     """Check IForest for various parameter settings on sparse input."""
-    rng = check_random_state(0)
+    rng = check_random_state(global_random_seed)
     X_train, X_test, y_train, y_test = train_test_split(
         diabetes.data[:50], diabetes.target[:50], random_state=rng
     )
```
```suggestion
    X_train, X_test = train_test_split(diabetes.data[:50], random_state=rng)
```

glemaitre commented on sklearn/ensemble/tests/test_iforest.py:
```diff
@@ -143,41 +143,41 @@ def test_max_samples_attribute():
-    rng = check_random_state(0)
+    rng = check_random_state(global_random_seed)
 
     X_train, X_test, y_train, y_test = train_test_split(
         diabetes.data, diabetes.target, random_state=rng
     )
```
```suggestion
    X_train, X_test, y_train, y_test = train_test_split(diabetes.data, random_state=rng)
```

glemaitre commented on sklearn/ensemble/tests/test_iforest.py:
```diff
@@ -143,41 +143,41 @@ def test_max_samples_attribute():
+    y_test = np.array([0] * 200 + [1] * 200)
 
     # fit the model
     clf = IsolationForest(max_samples=100, random_state=rng).fit(X_train)
```
You should add the global random state for the `test_iforest_works` test function.

MaxwellLZH replied:
updated :)

glemaitre reviewed (commented):
Otherwise, it looks good.

MaxwellLZH committed:
does not use global random state in test functions

MaxwellLZH committed:
clean up code in train_test_split

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

MaxwellLZH committed:
clean up code in train_test_split

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

MaxwellLZH committed:
clean up train_test_split & fix failing tests

glemaitre commented on sklearn/ensemble/tests/test_iforest.py:
```diff
@@ -44,7 +44,7 @@
 diabetes.target = diabetes.target[perm]
 
```
If you could still remove line 30 and locally create a random state in the function that requires one (or just pass an integer) it would be better.

MaxwellLZH replied:
Updated as suggested :)

MaxwellLZH committed:
remove global random vareiable

ogrisel committed:
Use the global_random_seed fixture also in test_iforest_chunks_works*

ogrisel committed:
[all random seeds]

ogrisel commented:
I pushed a new commit to also use the fixture in the `test_iforest_chunks_works*` tests and then re-triggered the `[all random seeds]` CI.

ogrisel reviewed (approved):
Assuming CI will stay green with the last commits, LGTM.

ogrisel committed:
[all random seeds]

test_iforest_sparse
test_iforest_parallel_regression
test_iforest_performance
test_iforest_chunks_works1
test_iforest_chunks_works2
